src/arch/xtensa/include/arch/task.h

```diff
@@ -46,7 +46,7 @@
 struct irq_task {
 	spinlock_t lock;	/**< lock */
 	struct list_item list;	/**< list of tasks */
-	uint32_t irq;		/**< IRQ level */
+	int irq;		/**< IRQ level */
```

Comment:
Why do we need to change `uint32_t`s to `int`s or `unsigned long`s? I know ATM it doesn't matter as we have only 32b archs, but it can get messy if we get 64b one time (or even if we start using some weird compiler for the same arch). I think types with guaranteed sizes are better.

Reply:
@jajanusz I strongly disagree. C is a high level strongly-typed language and it has types for different tasks, according to their semantics. Those types are free to be implemented differently on different arches *as long as* they satisfy certain requirements. E.g. int is at least 32 bits (don't ask me for a quote), etc.
Fixed-width types (uint32_t and friends) should be used only where *needed*. Where the width must be fixed. E.g. when implementing ABIs, where data can be exchanged between heterogeneous platforms. Or when copying data to and from hardware registers or buffers, which is also an ABI. "Normal" programs, using variables for internal purposes to implement program logic, have no need to fix variable width. You use "long" just because you want a "long" and you don't care that on some platforms it can be 32 and on others 64 bits. I actually *think*, that "long" is one of the few types that can have different length in gcc. "int" is 32 bits on both 32- and 64-bit platforms, "long long" is 64 bits, although it might be 128 bits on some. Anyway, where you need a counter, you don't need "(u)int32_t", you use "(unsigned) int" instead. And "irq" is now just a counter - it begins with low values for DSP native interrupts, then continues with values above 32 for cascaded interrupts, etc. I'm *quite* sure we'll never saturate 31 bits (for signed int) with our interrupt numbers. In fact I'd propose a patch (series) to replace many needlessly fixed-width types with normal C types.

Reply:
By writing that they are better I meant that in this case they are *easier/safer* to use. With fixed width you provide code that works with 1 width, with f.e. int you should give something that works with all sizes it can get. So it needs more caution, but if you know what you are doing then it's ok. We will have more *abstract* / *optimal* implementation.

These types have their uses, I just wanted to know in general what's your reasoning behind this.
You're mostly right regarding fixed widths, but I think *flags* variables may be better as fixed width variables.
Also I wasn't just worried about fixed widths. There are few places that change signed<-> unsigned and I wanted to be sure that's intended.

Reply:
@jajanusz yes, there are a few more cases for fixed-width: bitfields (though arguably, if we know we only need up to 32 bits, I'd be quite comfortable using unsigned int or unsigned long), ABIs between different languages (C / asm / java / ...), maybe sometimes forced alignment, things like that.
I'm sure "int" is perfectly sufficient for interrupt numbers now, that they aren't bitfields anymore. And yes, this patch set leaves a few functions, still using uint32_t for irq, many of them should be changed too. I can do that in a follow up patch.

Reply:
@lyakh we also need to be able to run SOF code on 64bit host processors as part of a heterogeneous pipelines, hence the need to align types to have similar and deterministic behaviour on both the host and DSP sides of a heterogeneous pipeline.

Reply:
@lgirdwood all the ABI data should have fixed width, but `for (i = 0; i < N; i++)` will work just perfectly deterministicly with `unsigned int i` on both 32- and 64-bit platforms, as long as N is within the range obviously.

Reply:
@lyakh loops are fine, but my main concern aside from ABI is the maths/logic used in processing algorithms.

Reply:
@lgirdwood sure, if you use (inline) assembly with some DSP maths instructions or similar, for that you might need to use very precise variable types. But irq numbers aren't used for that as far as I can see. Can you give any example where specifically you foresee problems?

Reply:
@lyakh I said I'm fine with loops and IRQ numbers as long as we are not changing types at runtime for no reason. I assume this is the case with this PR ? (and I've not reviewed yet, but will over the weekend). I do want to merge this soon so we can move on...

Reply:
@lgirdwood ok, sorry, misunderstood. As far as I can see, any type conversions like assignments between `int`, `unsigned int`, `int32_t`, `uint32_t` are safe. If you agree, later I can make a couple of cosmetic patches to replace at least some superfluous fixed-width types with standard C types.